src/java/org/apache/cassandra/index/sai/disk/vector/VectorMemtableIndex.java

```diff
@@ -196,25 +197,24 @@ public RangeIterator<PrimaryKey> search(QueryContext queryContext, Expression ex
 
         var keyQueue = graph.search(qv, limit, bits);
         if (keyQueue.isEmpty())
-            return RangeIterator.emptyKeys();
+            return RangeIterator.empty();
         return new ReorderingRangeIterator(keyQueue);
     }
 
     @Override
-    public RangeIterator<PrimaryKey> limitToTopResults(QueryContext context, RangeIterator<PrimaryKey> iterator, Expression exp, int limit)
+    public RangeIterator limitToTopResults(QueryContext context, List<PrimaryKey> keys, Expression exp, int limit)
     {
         if (minimumKey == null)
         {
             assert maximumKey == null : "Minimum key is null but maximum key is not";
-            return RangeIterator.emptyKeys();
-        }
-        Set<PrimaryKey> results = new HashSet<>();
-        while (iterator.hasNext())
-        {
-            var key = iterator.next();
-            if (!context.containsShadowedPrimaryKey(key))
-                results.add(key);
+            return RangeIterator.empty();
         }
+        // TODO revisit collection type. keys are already in correct order, so in brute force scenario, we are
```

Comment:
as mentioned above, I think you need two loops to fix this while feeding the correct size to brute force estimate

Reply:
I took care of this with https://github.com/datastax/cassandra/pull/758/commits/688c659ba097021606cbbd008d502262fa018be7. Since this is in the memtable, we don't have the same issue as above. The problem was unnecessarily using a HashSet, which lost track of an already ordered set.